### PR TITLE
Separate Delayed into DelayedFunction

### PR DESCRIPTION
**Pull request by mrocklin**
Added a `prefix=` keyword to delayed

    dfunc = delayed(function, prefix='add')

added a `dask_key_name=` keyword to calling a delayed function:

    dfunc(1, 2, dask_key_name='add-1-2')

It felt odd adding extra function-specific keyword arguments to the
`Delayed` class.  It feels like we're overloading one class with two
very different meanings so I'm also reseparating `Delayed` to
`DelayedFunction`

cc @jcrist

### CODE REVIEW

**Comment by jcrist** [comment]
The reason for the way I did it was that we already overload `__call__` for `Delayed` types - that's how methods work. Doing it this way means that callable objects no longer overload the same way. If you want to separate into two classes instead of using a flag, I'd separate like:

- `delayed(some_object) -> DelayedRoot` (with a better class name)
- `delayed(some_object)(a, b, c) -> Delayed`

The distinction is that the graph for `DelayedRoot` is the identity function, which can be optimized upon call (or other operations as well, if we wanted to get fancy). I'd make `DelayedRoot` subclass `Delayed` as well, and only overload `__call__` (and other things as necessary to make it work).

File: dask/delayed.py [hunk 1]
```diff
@@ -246,28 +249,68 @@ def _inner(self, other):
     return _inner
 
 
+class DelayedFunction(base.Base):
+    _optimize = staticmethod(lambda dsk, keys, **kwargs: dsk)
+    _finalize = staticmethod(first)
+    _default_get = staticmethod(get_sync)
```
Comment (jcrist) [inline dask/delayed.py]: Leftover from debugging?

Comment (mrocklin) [inline dask/delayed.py]: The graph will only ever have one element.  Single threaded seems best.

Comment (jcrist) [inline dask/delayed.py]: The way the generic `compute` works, `compute(delayed_function, delayed)` will barf, because they define two different schedulers as the default.

**Comment by mrocklin** [comment]
> Doing it this way means that callable objects no longer overload the same way

We call `delayed(apply)(self._key, args, kwargs)` which walks through the same codepath.

With the `DelayedRoot` approach would `delayed(add) + 1` not raise an error?  I was in favor of removing the `do/value` distinction for users but to me it seems like an odd thing to combine internally.

File: dask/delayed.py (hunk 1 above)
Comment (mrocklin) [inline dask/delayed.py]: Seems inevitable though, right?  Is threaded or multiprocessing more appropriate here?  I would claim that neither is particularly more accurate.  Might as well go with what makes more sense for this object and, in the rare case that this gets called, suffer the informative error.

**Comment by jcrist** [comment]
Sorry, I wasn't clear there. I meant that for a callable object, `delayed(obj)` will no longer overload operators, etc...

> With the `DelayedRoot` approach would delayed(add) + 1 not raise an error?

In the simplest implementation it wouldn't until runtime, since we'd treat the wrapped object as just a generic object. However, since we do know the type of the object in the root nodes, we could catch that at build time, since `getattr(add, '__add__')` would fail.

Basically, I'm just suggesting that you don't switch in `delayed` for callable/non-callable, and pass all things to `DelayedRoot`, which special cases `__call__` (at a minimum) to make the resulting graph simpler.

**Comment by mrocklin** [comment]
OK, that is a good point.  Will rethink.

**Comment by mrocklin** [comment]
OK, small change, not sure if it makes things more palatable. 

I'm sticking with `DelayedFunction` rather than `DelayedRoot`.  The extra attributes I care about all fall along the callable-root distinction rather than the root/not-root distinction.  The `DelayedFunction` class inherits from `Delayed` but has a few more bells and whistles around purity and naming.

File: dask/delayed.py [hunk 2]
```diff
@@ -208,11 +209,13 @@ def delayed(obj, name=None, pure=False):
     """
     if isinstance(obj, Delayed):
         return obj
-    task, dasks = to_task_dasks(obj)
-    root = not dasks
-    name = name or '%s-%s' % (type(obj).__name__, tokenize(task, pure=True))
-    dasks.append({name: task})
-    return Delayed(name, dasks, pure=pure, root=root)
+    if callable(obj):
+        return DelayedFunction(obj, pure=pure, prefix=prefix)
```
Comment (mrocklin) [inline dask/delayed.py]: Hrm, this is still a bit wonky though.

File: dask/delayed.py [hunk 3]
```diff
@@ -208,11 +209,15 @@ def delayed(obj, name=None, pure=False):
...
         return obj
+
     task, dasks = to_task_dasks(obj)
-    root = not dasks
-    name = name or '%s-%s' % (type(obj).__name__, tokenize(task, pure=True))
-    dasks.append({name: task})
-    return Delayed(name, dasks, pure=pure, root=root)
+
+    if not dasks and callable(obj):
+        return DelayedFunction(obj, pure=pure, prefix=prefix)
```
Comment (jcrist) [inline dask/delayed.py]: I'd feel a bit better if this was:

```python
if not dasks:
    return DelayedRoot(obj, pure=pure, prefix=prefix)    # Or some other class name
else:
    ....
```

No harm is done if the object being wrapped isn't callable, and the type of the output of delayed is easier to reason about (imo).

Comment (mrocklin) [inline dask/delayed.py]: If its not callable then it seems to me that that keywords/state like prefix and pure are no longer necessary and that `Delayed` is the better choice.

Comment (jcrist) [inline dask/delayed.py]: Perhaps, but there's still no harm in attaching them. I frequently do `delayed = delayed(pure=True)` at the top of a file. With this definition, `delayed(1)` should still be valid imo, so forbidding pure with non-callables isn't the best (and you don't do that here). I just find it easier to reason about if you can say "`delayed` on things with dask graphs returns `Delayed`, otherwise `DelayedRoot`".

Comment (mrocklin) [inline dask/delayed.py]: Can you explain your motivation there?  I don't understand the value behind the `DelayedRoot` abstraction.

Comment (jcrist) [inline dask/delayed.py]: Sure. I guess it's just a difference between how we divide the two uses of `delayed`. I see it as wrapping either a thing with dask graphs, resulting in a `Delayed`, or some immediate python object, resulting in a `DelayedRoot` (the root indicating it's at the base of the dag, there's probably a better name).

The special-casing of `__call__` for immediate functions is really just one of the things you can special case for any immediate object. For example, one could also do attribute/method checking at graph build time. So I don't see a distinction between callable/non-callable, rather I see one between lazy/immediate. And I find it easier to describe this dichotomy rather than one between callable objects that don't contain dasks/everything else.

As an aside, for the keyword thing perhaps drop `prefix` and just use `name`/part of name instead? Or is there a good reason to have these be separate?

Comment (mrocklin) [inline dask/delayed.py]: OK, the comparison with attribute checking makes sense to me.

Comment (mrocklin) [inline dask/delayed.py]: Thoughts on calling this a `DelayedLeaf`?  To me trees in CS grow down and leaves are often atoms in expressions.

Comment (jcrist) [inline dask/delayed.py]: Sounds good to me.

**Comment by mrocklin** [comment]
@jcrist this is ready for another (hopefully brief) review

File: dask/delayed.py [hunk 4]
```diff
@@ -6,8 +6,9 @@
 import operator
 import uuid
 
-from toolz import merge, unique, curry, first
+from toolz import merge, unique, curry, first, assoc
 
+from .async import get_sync
```
Comment (jcrist) [inline dask/delayed.py]: Both `get_sync` and `assoc` imports are added, but unused.

File: dask/base.py [hunk 5]
```diff
@@ -141,7 +142,14 @@ def normalize_function(func):
...
     else:
-        return str(func)
+        try:
+            return pickle.dumps(func, protocol=0)
+        except:
+            try:
+                import cloudpickle
+                return cloudpickle.dumps(func, protocol=0)
+            except:
+                return str(func)
```
Comment (jcrist) [inline dask/base.py]: This seems fine, I'm just curious about what caused this change.

File: dask/delayed.py (hunk 4 above)
Comment (mrocklin) [inline dask/delayed.py]: I've gone through with flake8 separately.  I'd like to do this as a different PR

**Comment by jcrist** [comment]
This looks pretty great. 

Note that both the docstring and prose docs still indicate that `delayed(x) -> Delayed` for all cases. This isn't exactly correct, but isn't wrong either (`isinstance(delayed(x), Delayed) == True`, and in the user's eyes really shouldn't matter). No need to change the docs, but also wouldn't be against a change. Up to you.

File: dask/base.py (hunk 5 above)
Comment (mrocklin) [inline dask/base.py]: Conversations with @maxhutch made it clear that functions weren't being tokenized consistently across Python sessions

```python
In [1]: from toolz import identity

In [2]: str(identity)
Out[2]: '<function identity at 0x7fac72f4a598>'
```

File: dask/delayed.py (hunk 4 above)
Comment (jcrist) [inline dask/delayed.py]: I just meant that you added them in this PR, but they're not used. Up to you.
